flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java

```diff
@@ -0,0 +1,291 @@
+		this.fatalErrorHandler = checkNotNull(fatalErrorHandler);
+		this.leaderContenderDescription = checkNotNull(leaderContenderDescription);
+
+		leaderLatch = new LeaderLatch(client, checkNotNull(latchPath));
+		cache = new NodeCache(client, leaderPath);
+
+		client.getUnhandledErrorListenable().addListener(this);
+
+		leaderLatch.addListener(this);
+		leaderLatch.start();
+
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(listener);
+
+		running = true;
+	}
+
+	@Override
+	public void close() throws Exception{
+		if (!running) {
+			return;
+		}
+		running = false;
+
+		LOG.info("Closing {}", this);
+
+		client.getUnhandledErrorListenable().removeListener(this);
+
+		client.getConnectionStateListenable().removeListener(listener);
+
+		Exception exception = null;
+
+		try {
+			cache.close();
+		} catch (Exception e) {
+			exception = e;
+		}
+
+		try {
+			leaderLatch.close();
+		} catch (Exception e) {
+			exception = ExceptionUtils.firstOrSuppressed(e, exception);
+		}
+
+		if (exception != null) {
+			throw new Exception("Could not properly stop the ZooKeeperLeaderElectionDriver.", exception);
+		}
+	}
+
+	@Override
+	public boolean hasLeadership() {
+		assert(running);
+		return leaderLatch.hasLeadership();
+	}
+
+	@Override
+	public void isLeader() {
+		leaderElectionEventHandler.onGrantLeadership();
+	}
+
+	@Override
+	public void notLeader() {
+		leaderElectionEventHandler.onRevokeLeadership();
+	}
+
+	@Override
+	public void nodeChanged() throws Exception {
+		if (leaderLatch.hasLeadership()) {
+			ChildData childData = cache.getCurrentData();
+			if (childData != null) {
+				final byte[] data = childData.getData();
+				if (data != null && data.length > 0) {
+					final ByteArrayInputStream bais = new ByteArrayInputStream(data);
+					final ObjectInputStream ois = new ObjectInputStream(bais);
+
+					final String leaderAddress = ois.readUTF();
+					final UUID leaderSessionID = (UUID) ois.readObject();
+
+					leaderElectionEventHandler.onLeaderInformationChange(
+						LeaderInformation.known(leaderSessionID, leaderAddress));
+					return;
+				}
+			}
+			leaderElectionEventHandler.onLeaderInformationChange(LeaderInformation.empty());
+		}
+	}
+
+	/**
+	 * Writes the current leader's address as well the given leader session ID to ZooKeeper.
+	 */
+	@Override
+	public void writeLeaderInformation(LeaderInformation leaderInformation) {
+		assert(running);
+		// this method does not have to be synchronized because the curator framework client
+		// is thread-safe. We do not write the empty data to ZooKeeper here. Because check-leadership-and-update
+		// is not a transactional operation. We may wrongly clear the data written by new leader.
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Write leader information: {}.", leaderInformation);
+		}
+		if (leaderInformation.equals(LeaderInformation.empty())) {
+			return;
```

Comment:
I think the normal behavior should be that once the current `LeaderElectionDriver` is not the leader, it is responsible for cleaning up the leader information on the external storage(Kubernetes, ZooKeeper). And it should only be executed when a new leader has not write his leader information.

For ZooKeeper, we are using an ephemeral node for leader information and it will be cleaned up automatically when ZooKeeper client timeout. So it is reasonable to not do the clean up here.

However, for Kubernetes, we should and could do this. Since check the leadership(aka check the annotation on the leader ConfigMap) and remove leader information is a transactional operation, we could guarantee that we will not overwrite a new leader information.

Reply:
I will add more description on the interface `LeaderElectionDriver#writeLeaderInformation`.

Reply:
The described behaviour sounds good to me @wangyang0918.